### PR TITLE
register new user

### PR DESCRIPTION
**Pull request by MohammedIrheem**
**Register New User**
  - validate user data by using "yup"
  - make queries to:
    - check if this user already exists or not 
    - add the user info to the database  
  - check if the event the user wants to book exists or not!
  - using the generating function to generate user code 
  - send SMS with validation code to the user

- before running this code you have to install NEXMO lib.
   - ' nmp i '

### CODE REVIEW

File: server/controller/routes/user/registeration.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+const registerValidation = (req, res, next) => {
+  const schema = yup.object().shape({
+    firstName: yup.string().required().max(10),
+    lastName: yup.string().required().max(10),
+    email: yup.string().email(),
+    mobile: yup.number().positive().integer().required(),
+    location: yup.string().required(),
+
+  });
+
```
Comment (AlaaTaima) [inline server/controller/routes/user/registeration.js]: **good work at validation :+1:**

File: server/controller/routes/user/registeration.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  }).catch(next);
+};
+
+// insert the valid user to the database
+const addUserToDB = (req, res, next) => {
+  newUser(req.body).then(() => res.json({ msg: 'Registerd successfuly' })).catch(next);
+};
+
+
+module.exports = { registerValidation, newUserExist, addUserToDB };
```
Comment (AlaaTaima) [inline server/controller/routes/user/registeration.js]: good work at error handling :+1:

File: server/router/index.js [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
-
+const {
+  serverError,
+  clientError,
+  registerValidation,
+  newUserExist,
+  addUserToDB,
+} = require('../controller');
+
+router.post('/register', registerValidation, newUserExist, addUserToDB);
```
Comment (AlaaTaima) [inline server/router/index.js]: good work at using related middle ware :+1:

File: server/controller/routes/user/registeration.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    mobile: yup.number().positive().integer().required(),
+    location: yup.string().required(),
+
+  });
+
+  schema.validate({
+    firstName: req.body.firstName,
+    lastName: req.body.lastName,
+    email: req.body.email,
+    mobile: req.body.mobile,
```
Comment (AlaaTaima) [inline server/controller/routes/user/registeration.js]: mobile number will be got from `req.params`.
edit it also in `newUser query` and `addUserToDB function` , they will take two parameters to get all data `(req.body , req.params) `

File: server/app.js [hunk 5]
```diff
@@ -20,6 +20,6 @@ app.use(express.static(join(__dirname, '..', 'client', 'build')));
 app.get('*', (req, res) => {
   res.sendFile(join(__dirname, '..', 'client', 'build', 'index.html'));
 });
-app.use(router);
+app.use('/api/v1', router);
```
Comment (AlaaTaima) [inline server/app.js]: `  app.use('/api/v1', router);`

```
 app.get('*', (req, res) => {
  res.sendFile(join(__dirname, '..', 'client', 'build', 'index.html'));
```

File: server/controller/routes/user/registerNewUser.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    email: req.body.email,
+    location: req.body.location,
+    mobile: req.body.mobile,
+
+  }, { abortEarly: false }).then(() => {
+    next();
+  }).catch((err) => {
+    res.status(400).json({ message: err.message });
+  });
+};
```
Comment (AhmedSafi97) [inline server/controller/routes/user/registerNewUser.js]: as long as you are using the same key names in the body request for your schema, you can pass the request body object directly  like this : 
```js
  schema.validate(req.body , { abortEarly: false }).then(() => {
    next();
  }).catch((err) => {
    res.status(400).json({ message: err.message });
  });
};
```

File: server/database/queries/users/newUser.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+const connection = require('../../config/connection');
+
+const newUser = (data) => connection.query({
+  text: 'INSERT INTO users(first_name, last_name, email, mobile, location) VALUES ($1, $2, $3, $4, $5)',
+  values: [data.firstName, data.lastName, data.email, data.mobile, data.location],
+});
```
Comment (AhmedSafi97) [inline server/database/queries/users/newUser.js]: I suggest using destructuring   here like this :
```js 
const newUser = ({firstName, lastName, email, mobile, location}) => connection.query({  text: 'INSERT 
INTO users(first_name, last_name, email, mobile, location) VALUES ($1, $2, $3, $4, $5)',
  values: [firstName, lastName, email, mobile, location],
});

File: server/controller/routes/user/sendSmsMsg.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+const Nexmo = require('nexmo');
+// Init Nexmo
+const nexmo = new Nexmo(
+  {
+    apiKey: '6bce011d',
+    apiSecret: 'C0RAsVGz6UvFWs9N',
+  },
+  { debug: true },
+);
```
Comment (AhmedSafi97) [inline server/controller/routes/user/sendSmsMsg.js]: api key and api secret must be hidden in the config file

File: server/controller/routes/user/sendSmsMsg.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+const sendSms = (req, res, next) => {
+  const { mobile, userCode } = req.body;
+  const text = `Hi, Your validation code is  ${userCode}`;
```
Comment (AhmedSafi97) [inline server/controller/routes/user/sendSmsMsg.js]: @AlaaTaima I suggest using more descriptive sms, what do you think ?

File: server/controller/routes/user/registerNewUser.js (hunk 6 above)
Comment (AlaaTaima) [reply server/controller/routes/user/registerNewUser.js]: good note :+1:

File: server/controller/routes/user/sendSmsMsg.js (hunk 9 above)
Comment (AlaaTaima) [reply server/controller/routes/user/sendSmsMsg.js]: yes , we should use more descriptive message , ex: 
" thanks for your registration, this your confirmation attendance code : 200 "

File: server/controller/routes/user/sendSmsMsg.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        error.status = 400;
+        next(error);
+      } else {
+        const data = {
+          id: responseData.messages[0]['message-id'],
+          number: responseData.messages[0].to,
+        };
+        // I've added the user number to the user to print the user message
+        // your validation code send to "user number"
+        req.user.number = data.number;
```
Comment (AlaaTaima) [inline server/controller/routes/user/sendSmsMsg.js]: good :+1:

File: server/router/index.js [hunk 11]
```diff
@@ -16,6 +21,17 @@ const {
 const { validationCancelReg } = require('../middleware');
 
 router.get('/getevents', getEvents);
+router.post('/register',
+  registerValidation,
+  newUserExist,
+  addUserToDB,
+  getUserID,
+  checkEventExist,
+  checkAlreadBooked,
```
Comment (AlaaTaima) [inline server/router/index.js]: there is no need for   `checkAlreadBooked`,  here we will deal with a new user that he hasn't  booked any event before.

File: server/controller/routes/user/registerNewUser.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  }).catch(next);
+};
+
+const getUserID = (req, res, next) => {
+  getUser(req.body.email).then(({ rows }) => {
+    const [user] = rows;
+    req.user = user;
+    next();
+  });
+};
```
Comment (AhmedSafi97) [inline server/controller/routes/user/registerNewUser.js]: getUserId is not necessary, you can return the row add using this syntax ```returning *``` after your insert query (addUserToDB)

Comment (AhmedSafi97) [inline server/controller/routes/user/registerNewUser.js]: getUserId is not necessary, you can return the row add using this syntax ```returning *``` after your insert query (addUserToDB)

File: server/router/index.js [hunk 13]
```diff
@@ -21,6 +26,16 @@ const {
...
+router.post('/register',
+  registerValidation,
+  newUserExist,
+  addUserToDB,
+  getUserID,
+  checkEventExist,
+  generateCode,
+  userWillAttend,
+  sendSms,
+  sendInvitation);
```
Comment (AhmedSafi97) [inline server/router/index.js]: Well, that's great, after a successful registration the event will be booked and an email and sms will be sent to the user, BUT it seems that you missed using the sms middle ware  after a user book an event in this route : 
``` js
router.post(
  '/checkUser',
  checkUser,
  checkEventExist,
  checkAlreadBooked,
  generateCode,
  userWillAttend,
  sendInvitation,
);
```

File: server/controller/routes/user/sendSmsMsg.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    text,
+    { type: 'unicode' },
+    (err, responseData) => {
+      if (err) {
+        const error = new Error();
+        error.msg = 'Invalid Number!';
+        error.status = 400;
+        next(error);
+      } else {
+        console.dir(responseData);
```
Comment (AlaaTaima) [inline server/controller/routes/user/sendSmsMsg.js]: remove console

**Pull request merged**